Libraries/LibJS/Runtime/ArrayIteratorPrototype.cpp

```diff
@@ -0,0 +1,92 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <LibJS/Runtime/Array.h>
+#include <LibJS/Runtime/ArrayIterator.h>
+#include <LibJS/Runtime/ArrayIteratorPrototype.h>
+#include <LibJS/Runtime/Error.h>
+#include <LibJS/Runtime/GlobalObject.h>
+#include <LibJS/Runtime/IteratorOperations.h>
+
+namespace JS {
+
+ArrayIteratorPrototype::ArrayIteratorPrototype(GlobalObject& global_object)
+    : Object(*global_object.iterator_prototype())
+{
+}
+
+void ArrayIteratorPrototype::initialize(Interpreter& interpreter, GlobalObject& global_object)
+{
+    Object::initialize(interpreter, global_object);
+    define_native_function("next", next, 0, Attribute::Writable | Attribute::Configurable);
+}
+
+ArrayIteratorPrototype::~ArrayIteratorPrototype()
+{
+}
+
+JS_DEFINE_NATIVE_FUNCTION(ArrayIteratorPrototype::next)
+{
+    auto this_value = interpreter.this_value(global_object);
+    if (!this_value.is_object() || !this_value.as_object().is_array_iterator_object())
+        return interpreter.throw_exception<TypeError>(ErrorType::NotAn, "Array Iterator");
+
+    auto& this_object = this_value.as_object();
+
+    auto& iterator = static_cast<ArrayIterator&>(this_object);
+    auto target_array = iterator.array();
+    if (target_array.is_undefined())
+        return create_iterator_result_object(interpreter, global_object, js_undefined(), true);
+    ASSERT(target_array.is_object());
+    auto& array = target_array.as_object();
+
+    auto index = iterator.index();
+    auto iteration_kind = iterator.iteration_kind();
+    // FIXME: Typed array check
+    auto length = array.indexed_properties().array_like_size();
+
+    if (index >= length) {
+        iterator.m_array = js_undefined();
+        return create_iterator_result_object(interpreter, global_object, js_undefined(), true);
+    }
+
+    iterator.m_index++;
+    if (iteration_kind == Object::PropertyKind::Key)
+        return create_iterator_result_object(interpreter, global_object, Value(static_cast<i32>(index)), false);
+
+    auto value = array.get(index);
+    if (interpreter.exception())
+        return {};
+    if (iteration_kind == Object::PropertyKind::Value)
+        return create_iterator_result_object(interpreter, global_object, value, false);
+
+    Array* entry_array = Array::create(global_object);
+    entry_array->put(0, Value(static_cast<i32>(index)));
+    entry_array->put(1, value);
```

Comment:
We should probably be more consistent with how we add values to an array - `indexed_properties().append()` vs `define_property()` vs `put()`. Feel free to decide which one is actually the best though! :smile:

https://github.com/SerenityOS/serenity/blob/ba641e97d9326be94b0b281fe140b5849c626b50/Libraries/LibWeb/Bindings/NavigatorObject.cpp#L45

https://github.com/SerenityOS/serenity/blob/7a1d485b19a43b3538b3ac69040ebc2b3bb14d85/Libraries/LibJS/Runtime/Object.cpp#L196-L203

Reply:
Great point. We should be using `define_property` when we've just created the object, since it avoids the prototype chain lookups and inserts the property directly into the object. When we're putting into an object we've received, we should use `put`. `indexed_properties.append` avoids a couple lines of code, but probably isn't worth it :)